### PR TITLE
Social Link: group all variations under one block type

### LINKED ISSUE #17277: Social Links Block: Allow plugins to add additional logos/social links
In #16897 a "social links" block is likely to land in its first version. 

Based on feedback for that first version, we'd like to explore some additional features. 

At launch, the Social Links block will have a wide range of icons for most present social networks and sites to link to. 

But there will always be some network, service, or tool that the block does not link to, but which the block does not contain. It would be good to explore how to add a simple way for plugins or themes to register new child blocks, so that the core list of blocks can remain relatively slim for maintenance purposes, but still allow flexibility for themers and developers.

### PR DESCRIPTION
**Pull request by mcsf**
Fixes #17277.

## Background

Initially, Social Links were defined (#16897) as a multitude of block types (`core/social-link-spotify`, `core/social-link-twitter`) all based on a mold of a single social link. The block types would allow only one parent, the container `core/social-links`. From a user standpoint this works, but there are downsides:
- In practice, this meant that the number of block types would grow with Social Links' number of supported sites.
- The document semantics arguably become too specialised, since `<!-- wp:social-link {"site":"wordpress","url":"https://example.org"} /-->` is descriptive and more generalisable.
- Because Social Links are currently defined as server-rendered blocks, there is overhead in registering all these variants on the server before page render. _Making Social Links static is a discussion to be had in #19888._
- From the previous point, it follows that REST API responses for the block type endpoints also increase in size to accommodate the generated block types.

## Description

Recently, the idea of Block Variations (#16283) has solidified (#19243). Initially oriented towards patterns of (inner) blocks, Variations can now accommodate more varied scenarios. One such scenario is that of Social Links.

This PR reimplements Social Links as made up one parent, `core/social-links`, and one child, `core/social-link`. The child then defines a number of variations, one per supported site. This represents no noticeable difference for the user, except for the change in the content markup:

Before|After
-|-
`<!-- wp:social-link-wordpress {"url":"https://example.org"} /-->`|`<!-- wp:social-link {"url":"https://example.org","site":"wordpress"} /-->`

## Screenshots 
social-links-block-variations

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. .

### CODE REVIEW

**Review by gziolo — COMMENTED** [review]
It looks great. Is there anything missing in the API?

I guess you should use store to get patterns to make this block extensible.

File: packages/block-library/src/social-link/index.js
Comment (mcsf) [inline packages/block-library/src/social-link/index.js]: I'm not very happy with this description, as the patterns offer many kinds of Web-based resources: social media profiles, multimedia content, misc. pages.

File: packages/block-library/src/social-link/index.php
Comment (mcsf) [inline packages/block-library/src/social-link/index.php]: I'd love to get rid of the loop above right away, in which all `core/social-link-<foo>` types are generated, but in order to do that we need a better plan for back-compat.

Right now only the editor has a back-compat provision via the change in `packages/blocks/src/api/parser.js`. How could we best handle this for server-side block rendering for the front-end?

File: packages/block-library/src/social-link/patterns.js
Comment (mcsf) [inline packages/block-library/src/social-link/patterns.js]: Unlike the Embed block, the Social Link block was initially conceived so that it can't be "generic", i.e. it has to follow a pattern. Given its purpose, this makes sense.

On the other hand, the way that Block Variations / Patterns are currently defined, if the "base" type is to be supplanted by the patterns there needs to be a default one. I chose _WordPress_ as the least opinionated one (or the least controversial opinion?).

I guess we could instead adapt the Patterns API so that it's possible to supplant the base type without requiring a default, but I'm not sure it's worth adding that compromise. What do you think, @gziolo and @mtias?

Comment (mcsf) [reply packages/block-library/src/social-link/patterns.js]: Side note: gotta love Vim for how magically it was able to transform `social-list.js` into `patterns.js` with just a quick macro. <3

**Comment by mcsf** [comment]
> I guess you should use store to get patterns to make this block extensible.

That's a good point. It's also something we may want to improve: shouldn't statically defined patterns be extensible or removable at runtime?

**Comment by mcsf** [comment]
> It looks great. Is there anything missing in the API?

Right now the only thing that comes to mind is https://github.com/WordPress/gutenberg/pull/19887#discussion_r371296884

**Comment by mkaz** [comment]
I'm wondering if Social Links should be just an Icon package for the logos, then the actual implementation of a social menu would look more like the Buttons block or Navigation block. This would do away with generating all of the individual site blocks.

Along the same lines, we could not have a special social links block and just add icon support to Buttons and/or Navigation.

**Comment by paaljoachim** [comment]
Marcus that sounds very interesting!
Having the social blocks use the Buttons Block implementation seems like a good idea.
In a sense there could be two almost identical blocks. Buttons and Social. The user might find it easier having a Buttons Block and a Social Block, even though they are almost identical.
@jasmussen

**Comment by jasmussen** [comment]
> Along the same lines, we could not have a special social links block and just add icon support to Buttons and/or Navigation.

Purely on a technical and code quality level, you all should do what you need to do! If navigation, social links and buttons are technically a single block, cool with me. It even makes sense insofar as one might very well want both navigation items AND a couple of social link icons in the very same navigation block.

However from a user and block library discoverability perspective, though, it is important that Navigation, Social Links and Buttons remain three separate blocks, though, so that you can search the block library for either and insert it with confidence.

Like I said, it could be like the embed block where the YouTube block is literally the same block as the Embed block, just with a different name and icon. But it's important that they each have a separate entry in the block library. Why? This goes all the way back to [the start](https://make.wordpress.org/core/2017/01/04/focus-tech-and-design-leads/):

> The editor will endeavour to create a new page and post building experience that makes writing rich posts effortless, and has "blocks" to make it easy what today might take shortcodes, custom HTML, **or "mystery meat" embed discovery**.

Emphasis mine.

**Comment by gziolo** [comment]
> However from a user and block library discoverability perspective, though, it is important that Navigation, Social Links and Buttons remain three separate blocks, though, so that you can search the block library for either and insert it with confidence.

This is exactly what we discussed on WordPress Slack with @mkaz and @mcsf yesterday (link requires registration):
https://wordpress.slack.com/archives/C02QB2JS7/p1580308195116200

We want to make it possible with new block variations (formerly patterns) API.

**Comment by mkaz** [comment]
For backward compatibility, can we add a deprecation to the social-links parent and use a migrate function which receives both attributes and InnerBlocks, and convert the InnerBlocks to the new social-link type. I think this can solve it in editor.

I'm not sure how we'd deprecate the server-side rendering

**Comment by mcsf** [comment]
> For backward compatibility, can we add a deprecation to the social-links parent and use a migrate function which receives both attributes and InnerBlocks, and convert the InnerBlocks to the new social-link type. I think this can solve it in editor.

This was solved early on by tackling the issue [right in the parser](https://github.com/WordPress/gutenberg/pull/19887/files#diff-7dae7b3890cf27659680c926b6992e31), following an established patterns of renamed blocks (e.g. `core/text`).

File: packages/block-library/src/social-link/edit.js
Comment (gziolo) [inline packages/block-library/src/social-link/edit.js]: Should it default to: `_( 'Social Icon' )` or use directly the name of the block?

```suggestion
	const socialLinkName = site ? getNameBySite( site ) : _( 'Social Icon' );
```

File: packages/block-library/src/social-link/index.js
Comment (gziolo) [inline packages/block-library/src/social-link/index.js]: Nit: now that it is only a single block, `block.json` can be used.

File: packages/block-library/src/social-link/index.php
Comment (gziolo) [inline packages/block-library/src/social-link/index.php]: With `block.json` in place, the following is possible:

https://github.com/WordPress/gutenberg/blob/2c587e5e252852870c087c1d9874d8867aa53c6d/packages/block-library/src/shortcode/index.php#L24-L34

It helps to avoid syncing attributes between PHP and JS. WordPress core has everything in place to handle it.

File: packages/block-library/src/social-link/social-list.js
Comment (gziolo) [inline packages/block-library/src/social-link/social-list.js]: It needs to be guarded in the case where a custom variation is registered and the site can't be found.

File: packages/block-library/src/social-link/variations.js
Comment (gziolo) [inline packages/block-library/src/social-link/variations.js]:
```suggestion
const variations = [
```
:)

File: packages/block-library/src/social-links/edit.js
Comment (gziolo) [inline packages/block-library/src/social-links/edit.js]: Nice :)

File: packages/blocks/src/api/parser.js
Comment (gziolo) [inline packages/blocks/src/api/parser.js]: Yes, this should be enough for now. In the long term, we should consider a filter that allows registering those handlers.  

```js
// Convert derivative blocks such as 'core/social-link-wordpress' to the
// canonical form 'core/social-link'.
const convertLegacySocialLink = ( { name, attributes } ) => {
    if ( name && name.indexOf( 'core/social-link-' ) === 0 ) {
        // Capture `social-link-wordpress` into `{"site":"wordpress"}`
        attributes.site = name.substring( 17 );
        name = 'core/social-link';
    }
    return { name, attributes };
}
```

You would register this type of filter just before the block gets registered.

**Review by gziolo — COMMENTED** [review]
I had some minor notes, but overall I like how it simplifies things. Code changes look good. It needs some testing, polishing and we should include it in WordPress 5.4 👍 

I also shared the initial proposal of how we could handle block name deprecations. It's something that should be explored separately though.

File: packages/block-library/src/social-link/edit.js
Comment (mcsf) [reply packages/block-library/src/social-link/edit.js]: That makes sense, or `Social Link` for consistency?

File: packages/block-library/src/social-link/index.php
Comment (mcsf) [reply packages/block-library/src/social-link/index.php]: True, thanks!

File: packages/block-library/src/social-link/variations.js
Comment (mcsf) [reply packages/block-library/src/social-link/variations.js]: 😬

File: packages/block-library/src/social-link/edit.js
Comment (mcsf) [reply packages/block-library/src/social-link/edit.js]: The title of the block itself is _Social Icon_, so I'll use that for now.

**Comment by mcsf** [comment]
Thanks for the great feedback, I've addressed everything.

**Comment by mtias** [comment]
I 100% agree with this direction. I don't think an abstraction such as "icon" or "buttons" would be a great idea, as it moves too far and loses semantically. One thing to consider is if `social-link` is better than `social-icon` for the purpose of this block.

File: packages/block-library/src/social-link/index.js
Comment (mtias) [inline packages/block-library/src/social-link/index.js]: This description could be improved. Reading it in isolation doesn't hint what the block does.

File: packages/block-library/src/social-links/edit.js
Comment (mtias) [inline packages/block-library/src/social-links/edit.js]: This looks a lot cleaner.

File: packages/block-library/src/social-link/block.json
Comment (mtias) [inline packages/block-library/src/social-link/block.json]: Is "site" the best attribute name?

File: packages/block-library/src/social-links/index.js
Comment (mtias) [inline packages/block-library/src/social-links/index.js]: Is this supposed to be "type" or "site"?

File: packages/block-library/src/social-link/index.js
Comment (mcsf) [reply packages/block-library/src/social-link/index.js]: Yes, same. My initial diff had a comment about it. Ideas?

File: packages/block-library/src/social-link/block.json
Comment (mtias) [reply packages/block-library/src/social-link/block.json]: What about `service`?

File: packages/block-library/src/social-links/index.js
Comment (mcsf) [reply packages/block-library/src/social-links/index.js]: Good catch.

File: packages/block-library/src/social-link/block.json
Comment (mcsf) [reply packages/block-library/src/social-link/block.json]: I've wondered about `service` too. Could try that.

File: packages/block-library/src/social-link/index.js
Comment (mkaz) [reply packages/block-library/src/social-link/index.js]: Create an icon link to external sites.

File: packages/block-library/src/social-link/block.json
Comment (mcsf) [reply packages/block-library/src/social-link/block.json]: How's this, @mtias? ba9070f

File: packages/block-library/src/social-link/index.js
Comment (mcsf) [reply packages/block-library/src/social-link/index.js]: I'll try _Display an icon linking to a social media profile or website_

**Comment by mkaz** [comment]
It looks like you're still updating the PR, but changing the attribute to `service` requires updates in edit.js and index.php for getting the icons to render properly. I don't mind the change, but concerned we're introducing another term, the code is already a mix of site, icon, and link.

**Comment by mkaz** [comment]
I pushed up an update to fix attribute change to service. A code quality improvement would probably be to change the function names to use service also. This can wait because I want to refactor how the SVG are displayed in #19888 and I might be changing functions.

**Review by mkaz — APPROVED** [review]
Great work :+1:

**Comment by mcsf** [comment]
> I pushed up an update to fix attribute change to service.

Thanks! Yes, this was a "commit-and-groceries" kind of dev flow. 😬 

> A code quality improvement would probably be to change the function names to use service also. This can wait because I want to refactor how the SVG are displayed in #19888 and I might be changing functions.

Agreed.

Thanks for the review!

**Comment by mcsf** [comment]
Before I merge, I pushed 07a9e36. Quoting the commit message:

> Remove unnecessary block deprecation
> 
> This PR introduces `core/social-link` as a new block type. Strictly speaking, any concerns of backwards compatibility should only deal with the migration of `core/social-link-FOO` types to this new one — something that the parser is handling directly, populating the new `service` attribute in that process.
> 
> Thus, no specific provisions are needed for the extant but obsolete `site` attribute: the attribute will be silently dropped during block validation. This can be attested by loading any content with links in the old `core/social-link-FOO` shape.

I'll wait for the tests to pass.

File: packages/block-library/src/social-link/index.php
Comment (gziolo) [reply packages/block-library/src/social-link/index.php]: You should check the HTML output of the edit post page. All blocks registered here will have a huge impact on the size of the page. In particular, the part that registers blocks to work with the `ServerSideRender` component (which is also pointless in this case as they will never be consumed).

At the same time, it's something that was never part of the WordPress core so we can provide the backward-compatibility only for the Gutenberg plugin and call it the day 😄

**Review by gziolo — COMMENTED** [review]
I haven't tested myself all this but code-wise it should be good to go once https://github.com/WordPress/gutenberg/pull/19887/files#r375784004 is addressed. It was one of the primary reasons why this block didn't land earlier in the WordPress core.

File: packages/block-library/src/social-link/block.json
Comment (gziolo) [inline packages/block-library/src/social-link/block.json]: I'm also wondering now that I noticed that the block's title is now `Social Icon`, should we change the name as well? :)

**Review by gziolo — APPROVED** [review]
I processed my previous comments and this is my conclusion, Let's land and improve block registration in PHP later. I don't see it as a blocker for the beta. In the plugin, it's going to work the same way anyway so we have a month to improve it :)

File: packages/block-library/src/social-link/block.json
Comment (mcsf) [reply packages/block-library/src/social-link/block.json]: When I started this PR I was pretty much on the `links` camp, but now I lean a bit towards `icon` too.

Comment (ZebulanStanphill) [reply packages/block-library/src/social-link/block.json]: @mcsf Why?

Comment (mcsf) [reply packages/block-library/src/social-link/block.json]: Because I was initially considering the primary function of these blocks to be the linking — and this fit in with the possibility of declaring Social Links as a variation of Navigation or Buttons. But now I think we may compromise on functionality if we center the semantics of these blocks around just the linking.

Instead, their real purpose lies more in the appearance: the rendering of the links as icons. In the future, the icons may change looks, colours or shape, but they will remain icons. I think the user is best served if we are clear about this. If, instead, the primary purpose is linking, then other blocks are better suited (Navigation, Buttons, File, etc.).

Comment (mcsf) [reply packages/block-library/src/social-link/block.json]: I have a diff that moves everything to Icons, but it's very large, so I'll open a separate PR to discuss this.

Comment (gziolo) [reply packages/block-library/src/social-link/block.json]: Sure thing 💯

Comment (ZebulanStanphill) [reply packages/block-library/src/social-link/block.json]: The point of having social icon links is that they're links, not that they're icons, isn't it? From an accessibility standpoint, I question whether we should even have a block dedicated to textless buttons in the first place. Wouldn't that be sort of an anti-pattern?

Also, what if I want to have social links with visible text labels? Would I use the Buttons block for that?

Comment (mcsf) [reply packages/block-library/src/social-link/block.json]: > The point of having social icon links is that they're links, not that they're icons, isn't it?

That's the subject part of the debate, I think.

> From an accessibility standpoint, I question whether we should even have a block dedicated to textless buttons in the first place. Wouldn't that be sort of an anti-pattern?

If done correctly, none of it should matter: icons should be as accessible as a regular text link, whether the reader cannot see, cannot use a cursor, etc.

> Also, what if I want to have social links with visible text labels? Would I use the Buttons block for that?

I could see either block type accommodating that use case.

I'll cc you on the new PR if you're interested in discussing this there.

File: packages/block-library/src/social-link/index.php
Comment (mcsf) [reply packages/block-library/src/social-link/index.php]: > At the same time, it's something that was never part of the WordPress core so we can provide the backward-compatibility only for the Gutenberg plugin and call it the day 😄

Yeah, that seems like the right way to go. Let's address in a subsequent PR.

Comment (mcsf) [reply packages/block-library/src/social-link/index.php]: #20098

**Pull request merged**
